Block/Service/AbstractCategoriesBlockService.php

```diff
@@ -0,0 +1,188 @@
+    {
+        $adminField = $this->getFormAdminType($formMapper, $this->categoryAdmin, 'categoryId', 'category', array(
+            'label' => 'form.label_category',
+        ), array(
+            'translation_domain' => 'SonataClassificationBundle',
+            'link_parameters' => array(
+                array(
+                    array(
+                        'context' => $block->getSetting('context'),
+                    ),
+                ),
+            ),
+        ));
+
+        $formMapper->add('settings', 'sonata_type_immutable_array', array(
+            'keys' => array(
+                array('title', 'text', array(
+                    'label' => 'form.label_title',
+                    'required' => false,
+                )),
+                array('context', 'choice', array(
+                    'label' => 'form.label_context',
+                    'required' => false,
+                    'choices' => $this->getContextChoices(),
+                )),
+                array($adminField, null, array()),
+            ),
+            'translation_domain' => 'SonataClassificationBundle',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureSettings(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults(array(
+            'title' => 'Categories',
+            'category' => false,
+            'categoryId' => null,
+            'context' => 'default',
+            'template' => 'SonataClassificationBundle:Block:base_block_categories.html.twig',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function load(BlockInterface $block)
+    {
+        if (is_numeric($block->getSetting('categoryId'))) {
+            $block->setSetting('categoryId', $this->getCategory($block->getSetting('categoryId')));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function prePersist(BlockInterface $block)
+    {
+        $this->resolveIds($block);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function preUpdate(BlockInterface $block)
+    {
+        $this->resolveIds($block);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getBlockMetadata($code = null)
+    {
+        $description = (!is_null($code) ? $code : $this->getName());
+
+        return new Metadata($this->getName(), $description, false, 'SonataClassificationBundle', array(
+            'class' => 'fa fa-folder-open-o',
+        ));
+    }
+
+    /**
+     * @param mixed $id
+     * @param mixed $default
+     *
+     * @return CategoryInterface
+     */
+    final protected function getCategory($id, $default = false)
+    {
+        if (is_object($id)) {
+            return $id;
+        }
+
+        if (is_numeric($id)) {
+            return $this->categoryManager->find($id);
+        }
+
+        return $default;
+    }
+
+    /**
+     * @param BlockInterface $block
+     */
+    private function resolveIds(BlockInterface $block)
```

Comment:
I disagree. If you "override" the method and it does something nasty then you don't necessarily get an error, but mess up your database. If you want to change some behaviour, you can change the lifecycle events, but this method should never be changed/overridden, etc, thus it should be final.

Making it final NOW doesn't hurt, as making it non-final later. It can be a problem in the other case.